src/backend.rs

```diff
@@ -233,202 +504,1527 @@ impl StackDepth {
+                        );
+                    }
+                }
+                ctx.block_state.regs.release_scratch_gpr(mem_ptr_reg);
+            }
+
+            assert!(offset <= i32::max_value() as u32);
+
+            let src = self.pop();
+            let base = self.pop();
+
+            let src_reg = self.into_reg(src);
+
+            match base {
+                ValueLocation::Immediate(i) => {
+                    let val = if let Some(i) = i.try_into() {
+                        Ok(i)
+                    } else {
+                        Err(self.into_temp_reg(base))
+                    };
+
+                    store_from_reg(self, src_reg, (offset as i32, val));
+
+                    if let Err(r) = val {
+                        self.block_state.regs.release_scratch_gpr(r);
+                    }
+                }
+                base => {
+                    let gpr = self.into_reg(base);
+                    store_from_reg(self, src_reg, (offset as i32, Err(gpr)));
+                    self.block_state.regs.release_scratch_gpr(gpr);
+                }
+            }
+
+            self.block_state.regs.release_scratch_gpr(src_reg);
+
+            Ok(())
+        }
+    }
 }
 
-pub fn literal_i32(ctx: &mut Context, imm: i32) {
-    let gpr = ctx.regs.take_scratch_gpr();
-    dynasm!(ctx.asm
-        ; mov Rd(gpr), imm
-    );
-    push_i32(ctx, gpr);
+trait TryInto<O> {
+    fn try_into(self) -> Option<O>;
 }
 
-pub fn relop_eq_i32(ctx: &mut Context) {
-    let right = pop_i32(ctx);
-    let left = pop_i32(ctx);
-    let result = ctx.regs.take_scratch_gpr();
-    dynasm!(ctx.asm
-        ; xor Rq(result), Rq(result)
-        ; cmp Rd(left), Rd(right)
-        ; sete Rb(result)
-    );
-    push_i32(ctx, result);
-    ctx.regs.release_scratch_gpr(left);
-    ctx.regs.release_scratch_gpr(right);
+impl TryInto<i64> for u64 {
+    fn try_into(self) -> Option<i64> {
+        let max = i64::max_value() as u64;
+
+        if self <= max {
+            Some(self as i64)
+        } else {
+            None
+        }
+    }
 }
 
-/// Pops i32 predicate and branches to the specified label
-/// if the predicate is equal to zero.
-pub fn pop_and_breq(ctx: &mut Context, label: Label) {
-    let predicate = pop_i32(ctx);
-    dynasm!(ctx.asm
-        ; test Rd(predicate), Rd(predicate)
-        ; je =>label.0
-    );
-    ctx.regs.release_scratch_gpr(predicate);
+impl TryInto<i32> for i64 {
+    fn try_into(self) -> Option<i32> {
+        let min = i32::min_value() as i64;
+        let max = i32::max_value() as i64;
+
+        if self >= min && self <= max {
+            Some(self as i32)
+        } else {
+            None
+        }
+    }
 }
 
-/// Branch unconditionally to the specified label.
-pub fn br(ctx: &mut Context, label: Label) {
-    dynasm!(ctx.asm
-        ; jmp =>label.0
-    );
+#[derive(Debug, Clone)]
+pub struct VirtualCallingConvention {
```

Comment:
No, it's just a really poor choice of name (it actually derives from a previous version of the code where I just had `VirtualStack` but then I added `StackDepth` too). This actually represents a calling convention that doesn't specify physical locations for its arguments, and so can have arguments in duplicated locations or constants for arguments. This is for blocks that we are certain have precisely one caller. For example, for a loop header we always have to allocate unique locations for each argument since we don't know whether two arguments that are in overlapping slots at the time that we're translating the first call will always be called with overlapping arguments.

A compiler that produces Microwasm directly could avoid this problem, since if a block is always called with overlapping arguments it can simply take fewer arguments and use `pick` at the beginning of the block, likewise if one of the arguments is always `const` then it could simply emit the `const` inside the block itself. This is really only a problem when we're translating from Wasm.